scripts/public_api_report.py

```diff
@@ -0,0 +1,795 @@
+            filepath=pathlib.Path(
+                "great_expectations/datasource/batch_kwargs_generator/batch_kwargs_generator.py"
+            ),
+        ),
+    ]
+
+    def __init__(
+        self,
+        repo_root: pathlib.Path,
+        docs_example_parser: DocsExampleParser,
+        code_parser: CodeParser,
+        public_api_checker: PublicAPIChecker,
+        excludes: Union[List[IncludeExcludeDefinition], None] = None,
+        includes: Union[List[IncludeExcludeDefinition], None] = None,
+    ) -> None:
+        """Create a CodeReferenceFilter.
+
+        Args:
+            repo_root: Repository root directory, for use in creating relative paths.
+            docs_example_parser: A DocsExampleParser initialized with the file
+                contents from all docs examples to process.
+            code_parser: A CodeParser initialized with library code.
+            public_api_checker: A PublicAPIChecker to aid in filtering.
+            excludes: Override default excludes by supplying a list of
+                IncludeExcludeDefinition instances.
+            includes: Override default includes by supplying a list of
+                IncludeExcludeDefinition instances. Note: Includes override
+                excludes if they are conflicting.
+        """
+        self.repo_root = repo_root
+        self.docs_example_parser = docs_example_parser
+        self.code_parser = code_parser
+        self.public_api_checker = public_api_checker
+
+        if not excludes:
+            self.excludes = self.DEFAULT_EXCLUDES
+        else:
+            self.excludes = excludes
+
+        if not includes:
+            self.includes = self.DEFAULT_INCLUDES
+        else:
+            self.includes = includes
+
+    def filter_definitions(self) -> Set[Definition]:
+        """Main method to perform all filtering.
+
+        Filters Definitions of entities (class, method and function).
+        Returned Definitions:
+            1. Appear in published documentation examples.
+            2. Are not private.
+            3. Are included or not excluded by an IncludeExcludeDefinition.
+            4. Are not marked with the @public_api decorator.
+
+        Returns:
+            Definitions that pass all filters.
+        """
+        usages_in_docs_examples: Set[str] = self._docs_examples_usages()
+        gx_definitions_used_in_docs_examples: Set[
+            Definition
+        ] = self._filter_gx_definitions_from_docs_examples(
+            gx_usages_in_docs_examples=usages_in_docs_examples
+        )
+        non_private_definitions: Set[Definition] = self._filter_private_entities(
+            definitions=gx_definitions_used_in_docs_examples
+        )
+        included_definitions: Set[Definition] = self._filter_or_include(
+            definitions=non_private_definitions
+        )
+        definitions_not_marked_public_api: Set[
+            Definition
+        ] = self._filter_for_definitions_not_marked_public_api(
```

Comment:
those marked as `public_api` will be addressed by linker (ie. DocString has correct format and appropriate components)